source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,659 @@
+    another function. Or sometimes it is necessary to introduce a few
+    variations, while carrying all the other aspects. Because of function
+    attributes, parameter storage classes and user-defined attributes, this
+    requires building a string mixin. In addition, the mixed-in code must refer
+    only to local names, if it is to work across module boundaires.
+
+    This module facilitates the creation of such mixins.
+
+ Copyright: Copyright Jean-Louis Leroy 2017-2020
+
+ License:   $(LINK2 http://boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+ Authors:   Jean-Louis Leroy
++/
+
+module bolts.experimental.refraction;
+
+import std.array;
+import std.format;
+import std.meta;
+import std.algorithm.iteration : map;
+import std.range : iota;
+import std.traits;
+
+template ParameterAttribute(alias F, int i, int j)
+{
+  static if (is(typeof(F) P == __parameters)) {
+    alias ParameterAttribute = Alias!(__traits(getAttributes, P[i..i+1])[j]);
+  }
+}
+
+/**
+    Return a Function object that captures all the aspects of Fun, using the
+    value of localSymbol to represent the return and parameter types, and the
+    UDAs.
+
+    Params:
+    fun = a function
+    localSymbol = a string mixin that represents fun in the caller's context
+
+*/
+
+Function refract(alias fun, string localSymbol)()
+  if (is(typeof(fun) == function))
+{
+  Function model = {
+  name: __traits(identifier, fun),
+  localSymbol: localSymbol,
+  returnType: "std.traits.ReturnType!("~localSymbol~")",
+  parameters: refractParameterList!(fun, localSymbol),
+  udas: __traits(getAttributes, fun)
+  .length.iota.map!(
+    formatIndex!("@(__traits(getAttributes, %s)[%%d])".format(localSymbol))).array,
+  attributes: functionAttributes!(fun),
+  static_: __traits(isStaticFunction, fun) && isAggregate!(__traits(parent, fun)),
+  body_: ";",
+  };
+
+  return model;
+}
+
+private enum isScope(alias T) = is(T == module) || is(T == struct)
+  || is(T == class) || is(T == interface);
+
+Function refract(
+  alias Scope, string localSymbol, string name, uint index)()
+  if (isScope!Scope)
+{
+  return refract!(
+    __traits(getOverloads, Scope, name)[index],
+    `__traits(getOverloads, %s, "%s")[%d]`.format(localSymbol, name, index))
+    .setIndex(index);
+}
+
+private enum True(T...) = true;
+
+Function[] refract(alias Scope, string localSymbol, alias Predicate = True)()
+  if (isScope!Scope)
+{
+  Function[] functions;
+
+  static foreach (member; __traits(allMembers, Scope)) {
+    static foreach (index, fun; __traits(getOverloads, Scope, member)) {
+      static if (Predicate!fun) {
+        functions ~= refract!(Scope, localSymbol, member, index);
+      }
+    }
+  }
+
+  return functions;
+}
+
+///
+unittest
+{
+  pure @nogc int answer(lazy string question);
+  alias F = answer; // typically F is a template argument
+  static assert(
+    refract!(F, "F").mixture ==
+    "pure @nogc @system std.traits.ReturnType!(F) answer(lazy std.traits.Parameters!(F)[0] _0);");
```

Comment:
> Is it possible to have the mixture have type `string` as the parameter type instead of `std.traits.Parameters!(F)[0]`? Maybe with some `bolts.StringOf` or `T.stringof` magic?

No, that would break everything. Well not if the type is `string`, because that exists in all scopes; but it would break for almost any user defined type (including those from Phobos).